File: app/src/main/java/br/com/jonathanzanella/myexpenses/views/DateTimeView.java
package br.com.jonathanzanella.myexpenses.views;

import android.app.DatePickerDialog;
import android.app.TimePickerDialog;
import android.content.Context;
import android.util.AttributeSet;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.TimePicker;

import org.joda.time.DateTime;

import java.text.SimpleDateFormat;
import java.util.Locale;

import br.com.jonathanzanella.myexpenses.R;
import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;
import lombok.Getter;
import lombok.Setter;

public class DateTimeView extends BaseView {
	public interface Listener {
		void onDateTimeChanged(DateTime currentTime);
	}

	public static final SimpleDateFormat dateFormat = new SimpleDateFormat("dd/MM/yy", Locale.getDefault());
	public static final SimpleDateFormat hourFormat = new SimpleDateFormat("HH:mm", Locale.getDefault());
	@Bind(R.id.view_datetime_date)
	EditText date;
	@Bind(R.id.view_datetime_hour)
	EditText hour;

	@Setter
	private Listener listener;

	@Getter
	private DateTime currentTime;

	public DateTimeView(Context context) {
		super(context);
	}

	public DateTimeView(Context context, AttributeSet attrs) {
		super(context, attrs);
	}

	public DateTimeView(Context context, AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);
	}

	@Override
	protected void init() {
		inflate(getContext(), R.layout.view_datetime, this);
		ButterKnife.bind(this);
	}

	public void setDate(DateTime currentTime) {
		this.currentTime = currentTime;
		onDateChanged();
	}

	private void onDateChanged() {
		date.setText(dateFormat.format(currentTime.toDate()));
		hour.setText(hourFormat.format(currentTime.toDate()));
		if(listener != null)
			listener.onDateTimeChanged(currentTime);
	}

	@OnClick(R.id.view_datetime_date)
	void onDate() {
		new DatePickerDialog(getContext(), new DatePickerDialog.OnDateSetListener() {
			@Override
			public void onDateSet(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
				currentTime = currentTime.withYear(year).withMonthOfYear(monthOfYear + 1).withDayOfMonth(dayOfMonth);
				onDateChanged();
			}
		}, currentTime.getYear(), currentTime.getMonthOfYear() - 1, currentTime.getDayOfMonth()).show();
	}

	@OnClick(R.id.view_datetime_hour)
	void onHour() {
		new TimePickerDialog(getContext(), new TimePickerDialog.OnTimeSetListener() {
			@Override
			public void onTimeSet(TimePicker view, int hourOfDay, int minute) {
				currentTime = currentTime.withHourOfDay(hourOfDay).withMinuteOfHour(minute);
				onDateChanged();
			}
		}, currentTime.getHourOfDay(), currentTime.getMinuteOfHour(), true).show();
	}
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/views/DateTimeView.kt
package br.com.jonathanzanella.myexpenses.views

import android.app.DatePickerDialog
import android.app.TimePickerDialog
import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.FrameLayout
import br.com.jonathanzanella.myexpenses.R
import kotlinx.android.synthetic.main.view_datetime.view.*
import org.joda.time.DateTime
import java.text.SimpleDateFormat
import java.util.*

class DateTimeView@JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr) {
    val dateFormat = SimpleDateFormat("dd/MM/yy", Locale.getDefault())
    val hourFormat = SimpleDateFormat("HH:mm", Locale.getDefault())

    private var listener: Listener? = null

    var currentTime: DateTime? = null
        private set

    interface Listener {
        fun onDateTimeChanged(currentTime: DateTime)
    }

    init {
        View.inflate(context, R.layout.view_datetime, this)
        date.setOnClickListener { onDate() }
        hour.setOnClickListener { onHour() }
    }

    fun setDate(currentTime: DateTime) {
        this.currentTime = currentTime
        onDateChanged()
    }

    private fun onDateChanged() {
        val time = currentTime!!
        date.setText(dateFormat.format(time.toDate()))
        hour.setText(hourFormat.format(time.toDate()))
        listener?.onDateTimeChanged(time)
    }

    internal fun onDate() {
        val time = currentTime!!
        DatePickerDialog(context, DatePickerDialog.OnDateSetListener { _, year, monthOfYear, dayOfMonth ->
            currentTime = currentTime?.withYear(year)?.withMonthOfYear(monthOfYear + 1)?.withDayOfMonth(dayOfMonth)
            onDateChanged()
        }, time.year, time.monthOfYear - 1, time.dayOfMonth).show()
    }

    internal fun onHour() {
        val time = currentTime!!
        TimePickerDialog(context, TimePickerDialog.OnTimeSetListener { _, hourOfDay, minute ->
            currentTime = currentTime?.withHourOfDay(hourOfDay)?.withMinuteOfHour(minute)
            onDateChanged()
        }, time.hourOfDay, time.minuteOfHour, true).show()
    }

    fun setListener(listener: Listener) {
        this.listener = listener
    }
}